File: norumanga-client/lib/data/models/payment.dart
/// A payment transaction record (the user's transaction history).
class Payment {
  final String id;
  final String accountId;
  final String bundleId;
  final String bundleName;
  final int amount; // VND
  final String method; // MOMO
  final String status; // PENDING / SUCCESS / FAILED
  final String? transactionRef;
  final DateTime? expiresAt;
  final DateTime? createdAt;

  Payment({
    required this.id,
    required this.accountId,
    required this.bundleId,
    required this.bundleName,
    required this.amount,
    required this.method,
    required this.status,
    this.transactionRef,
    this.expiresAt,
    this.createdAt,
  });

  factory Payment.fromJson(Map<String, dynamic> json) {
    return Payment(
      id: json['id'] as String? ?? '',
      accountId: json['accountId'] as String? ?? '',
      bundleId: json['bundleId'] as String? ?? '',
      bundleName: json['bundleName'] as String? ?? '',
      amount: (json['amount'] as num?)?.toInt() ?? 0,
      method: json['method'] as String? ?? 'MOMO',
      status: json['status'] as String? ?? 'PENDING',
      transactionRef: json['transactionRef'] as String?,
      expiresAt: _parseDate(json['expiresAt']),
      createdAt: _parseDate(json['createdAt']),
    );
  }

  bool get isSuccess => status == 'SUCCESS';

  static DateTime? _parseDate(dynamic value) {
    if (value == null) return null;
    try {
      return DateTime.parse(value.toString()).toLocal();
    } catch (_) {
      return null;
    }
  }
}

File: norumanga-server/src/main/java/prm/prmbackend/entity/Payment.java
package prm.prmbackend.entity;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;
import prm.prmbackend.entity.enums.PaymentMethod;
import prm.prmbackend.entity.enums.PaymentStatus;

import java.time.Instant;

/**
 * MongoDB collection: payments
 *
 * A transaction record for a user purchasing a subscription bundle.
 * Acts as the user's transaction history (read-only for end users).
 */
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Document(collection = "payments")
public class Payment {

    @Id
    private String id;

    /** Account id of the purchasing user */
    private String accountId;

    /** Bundle that was purchased */
    private String bundleId;

    /** Snapshot of bundle name at purchase time */
    private String bundleName;

    /** Amount charged in VND (snapshot of bundle price) */
    private long amount;

    private PaymentMethod method;

    private PaymentStatus status;

    /** External transaction reference (e.g. MoMo transaction id) */
    private String transactionRef;

    /** When premium access (granted by this payment) expires */
    private Instant expiresAt;

    private Instant createdAt;

    private Instant updatedAt;
}
